### PR TITLE
ECS Task IAM profile credentials ignored in repository-s3 plugin

### LINKED ISSUE #26913: ECS Task IAM profile credentials ignored in repository-s3 plugin
**Elasticsearch version** (`bin/elasticsearch --version`): 5.6.2

**Plugins installed**: [repository-s3]

**JVM version** (`java -version`): Doesn't seem to matter

**OS version** (`uname -a` if on a Unix-like system): Doesn't seem to matter

**Description of the problem including expected versus actual behavior**:

When using the repository-s3 plugin from within an ECS Task that has an [ECS Task IAM role
](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html), the repository-s3 instead loads the underlying ec2 host's credentials, which appears to be contrary (or at least unexpected!), based on what is in the [usage guide](https://www.elastic.co/guide/en/elasticsearch/plugins/5.6/repository-s3-usage.html)

**Steps to reproduce**:

 1. Install elasticsearch 5.6.2 on a docker container
 2. Use docker container in a task definition on AWS ECS
 3. Give task definition an IAM role with access to an s3 bucket
 4. Give EC2 instance an IAM role that denies access to s3 bucket
 5. Install repository-s3 plugin using `elasticsearch-plugin install repository-s3`
 6. Start elasticsearch
 7. Attempt to create a repository using as specified at [https://www.elastic.co/guide/en/elasticsearch/plugins/5.6/repository-s3-repository.html#repository-s3-repository](https://www.elastic.co/guide/en/elasticsearch/plugins/5.6/repository-s3-repository.html#repository-s3-repository), which fails.
 8. Note that the aws cli is, however able to access the s3 bucket from inside the container correctly (e.g., aws s3 ls s3::/<BUCKET_NAME> works).

**Believed cause**:

I believe I've tracked the issue down to https://github.com/elastic/elasticsearch/blob/16431a660188a7fb799de12df22a3193fbb491ec/plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/InternalAwsS3Service.java#L172, which uses an [InstanceProfileCredentialsProvider](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/InstanceProfileCredentialsProvider.html), rather than the [DefaultAWSCredentialsProviderChain](http://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/DefaultAWSCredentialsProviderChain.html), which I would have thought would be more appropriate, unless I am missing something?

### PR DESCRIPTION
**Pull request by vladimirdolzhenko**
ECS Task IAM profile credentials ignored in repository-s3 plugin

Closes #26913

Based on PR #31918 

`EC2ContainerCredentialsProviderWrapper` - utilises IAM role for a task - both relative to `http://169.254.170.2` and absolute + fail over to `InstanceProfileCredentialsProvider` that is currently in use. 

provider relies on **environment variables** common to containers

### CODE REVIEW

**Comment by DaveCTurner** [comment]
We discussed this and, as per https://github.com/elastic/elasticsearch/issues/26913#issuecomment-335394135, think that the `DefaultAWSCredentialsProviderChain` is too liberal with its choice of sources for credentials. However `com.amazonaws.auth.EC2ContainerCredentialsProviderWrapper` looks suitable: it only queries the metadata services, but supports both container credentials and instance profile credentials.

However I think we shouldn't proceed here without a bit more testing: specifically, that we do correctly get the credentials from the metadata services. I think we can extend `S3Fixture` to act as the appropriate metadata services: we can set the URI to the container metadata service using the `AWS_CONTAINER_CREDENTIALS_FULL_URI` environment variable, and the URI for the instance metadata service via the `com.amazonaws.sdk.ec2MetadataServiceEndpointOverride` system property.

**Comment by vladimirdolzhenko** [comment]
@DaveCTurner @atorok this PR extends #31918 - **environment** variable has to be specified to redirect ECS to custom endpoint

File: plugins/repository-s3/build.gradle [hunk 1]
```diff
@@ -302,8 +320,13 @@ task s3Fixture(type: AntFixture) {
   env 'S3FIXTURE_TEMPORARY_KEY', "${s3TemporaryAccessKey}"
   env 'S3FIXTURE_TEMPORARY_SESSION_TOKEN', "${s3TemporarySessionToken}"
   env 'S3FIXTURE_EC2_BUCKET_NAME', "${s3EC2Bucket}"
+  env 'S3FIXTURE_EC2_CREDENTIALS', "${s3EC2Credentials}"
```
Comment (atorok) [inline plugins/repository-s3/build.gradle]: There's no need for a Gstring here, could have used the  `s3EC2Credentials` variable directly.

**Review by alpar-t — DISMISSED** [review]
Gradle LGTM

File: plugins/repository-s3/build.gradle (hunk 1 above)
Comment (vladimirdolzhenko) [reply plugins/repository-s3/build.gradle]: good spot - thanks @atorok 👍

**Comment by vladimirdolzhenko** [comment]
@DaveCTurner could you pls have a look - it is ready for review as #31918 is merged to master

**Comment by DaveCTurner** [comment]
This seems to have been rebased, losing earlier review comments. Please could you try and avoid rebasing open PRs?

The changes to the `S3Service` and `AmazonS3Fixture` LGTM, but I'm unsure if there's a neater way to run the integ test clusters than this. @atorok or @ywelsch does this look ok to you?

**Comment by vladimirdolzhenko** [comment]
@DaveCTurner sorry for the rebase - it was really huge number of changes that this PR is based on

**Comment by atorok** [comment]
LGTM I don't think we can do better for now.

**Review by DaveCTurner — APPROVED** [review]
Ok thanks @atorok, LGTM too then.

**Comment by vladimirdolzhenko** [comment]
thanks @DaveCTurner and @atorok for the review

**Pull request merged**